[Ingest Pipelines] Processor Editor Move Tooltip

jloleysens opened the pull request:
## Summary

Adds a tooltip to ingest processors editor UI to indicate which processor is being moved. See images below.

## How to test

1. Start Kibana with default/basic license
2. Create the following pipeline https://gist.github.com/jloleysens/e9c92d0f26e16af2ea301fa7b6349815
3. Go to the ingest pipelines plugin in the management section. The first processor should have a description
4. Click the move button on the first processor
5. Scroll to the bottom of the list and place the processor last

Throughout these steps, when the tooltip is visible, it should stay roughly 20px below and 20px to the right of the cursor (even while scrolling).

## GIF

![tooltip](https://user-images.githubusercontent.com/8155004/87764148-327b5280-c816-11ea-898d-4ea9c18f7ecf.gif)

## Screenshots

**Move processor with description**
<img width="1036" alt="Screenshot 2020-07-17 at 10 10 22" src="https://user-images.githubusercontent.com/8155004/87764172-3effab00-c816-11ea-89af-b5c5b1ac9b62.png">

**Move processor without description**
<img width="740" alt="Screenshot 2020-07-17 at 10 10 33" src="https://user-images.githubusercontent.com/8155004/87764220-4e7ef400-c816-11ea-92d3-4938a0577395.png">

### Checklist

- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/)
* Commit: 47484b0e90fc714a1980a38e0734228603cf96dd

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/47484b0e90fc714a1980a38e0734228603cf96dd) | diff | [baseline](https://github.com/elastic/kibana/commit/35e1dd65228562b0db746b377969a45b5c7e6a56) |
| --- | --- | --- | --- |
| ingestPipelines | 267 | +9 | 258 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/47484b0e90fc714a1980a38e0734228603cf96dd) | diff | [baseline](https://github.com/elastic/kibana/commit/35e1dd65228562b0db746b377969a45b5c7e6a56) |
| --- | --- | --- | --- |
| ingestPipelines | 587.2KB | +7.7KB | 579.6KB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/","number":"62558","commit":"47484b0e90fc714a1980a38e0734228603cf96dd"}]}
PIPELINE-->

cjcenizal commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/components/pipeline_processors_editor_item/pipeline_processors_editor_item.tsx:
```diff
@@ -94,7 +94,13 @@ export const PipelineProcessorsEditorItem: FunctionComponent<Props> = memo(
           aria-label={label}
-          onChange={() => (!isMovingThisProcessor ? onMove() : onCancelMove())}
+          onChange={() => {
+            if (!isMovingThisProcessor) {
```
Nit: I find it easier to scan a condition if we're asserting against a positive statement, e.g. `if (isMovingThisProcessor)`. Seems like a nice readability win for a very small change.

cjcenizal commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/components/pipeline_processors_editor_item_tooltip/pipeline_processors_editor_item_tooltip.tsx:
```diff
@@ -0,0 +1,75 @@
+     * that we can render it anywhere the cursor can go.
+     */
+    <EuiPortal
+      insert={{
```
Why do we need this? I removed it locally and the tooltip seemed to behave as expected.

jloleysens replied:
Ah it looks like I misread the EuiPortal docs, the default behaviour appends to the body already:

https://github.com/elastic/eui/blob/f9b864a4695cfe5bfe8b79219dcb535361419bf3/src/components/portal/portal.tsx#L62

jloleysens commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/)
* Commit: 62a776642a632cfd6a8b26e7ec51036c8254767a

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/62a776642a632cfd6a8b26e7ec51036c8254767a) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 267 | +9 | 258 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/62a776642a632cfd6a8b26e7ec51036c8254767a) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 585.3KB | +7.7KB | 577.6KB |

### History
* :green_heart: [Build #62558](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/) succeeded 47484b0e90fc714a1980a38e0734228603cf96dd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/","number":"62558","commit":"47484b0e90fc714a1980a38e0734228603cf96dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/","number":"62870","commit":"62a776642a632cfd6a8b26e7ec51036c8254767a"}]}
PIPELINE-->

jloleysens commented:
> Component suggestion: EuiMouseTip

👆🏻 this is something I plan on discussing with the EUI team for sure, I'll be opening an issue on their repo soon! Thanks for weighing in with your thoughts, I'll be sure to factor them in to our discussion (and ping you as needed!)

> Component suggestion: EuiInlineEditable

This is something I had not thought about as much but I think certainly deserves some mention EUI-side too. This certainly merits a separate discussion/issue on EUI.

> Bug: Moving to top or bottom of processors list

I'm not sure I was able to replicate this to the extent that I'd view it as a bug. That being said, with the work identified in the a11y issue (https://github.com/elastic/kibana/issues/71666) we will be swapping out `react-virtualized` for `react-window`. This might influence the behaviour you have identified - we can review it then again.

> Bug: Moving while processor panel is open

Great catch! This was a particularly weird one because I was passing in `disabled` to `EuiButtonToggle` which was correctly updating the style to _look_ disabled. But at a functional level the button was still clickable. I fixed this by using the correct EUI API `isDisabled` and added a test for this too.

jloleysens commented:
> UX suggestion: Moving between regular and failure processors

We can definitely create an issue for tracking this! It is a pre-existing issue. Would you mind opening one?

> UX suggestion: "Can't move here" feedback

Thanks for sharing your insight here! I can see how the `not-allowed` cursor is perhaps adding unnecessary noise - removed it :)

As for the grey line and "cannot move here" tooltip; I think your point about noise is valid but only for users that are familiar with the functionality. In those cases I don't even think those users will see the "cannot move here" UI appearing because they will move over it too fast (from what I have observed). That leaves users who are new to this UI - for them I think the extra clear indication about entering "move" mode and introducing the drop zones off the bat is still valuable for discoverability.

For those reasons I'd prefer to punt on this point a little bit longer.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/)
* Commit: 8cfbe189cfe29b8d94f5600cfe1be58956777ca8
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute kibana-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/execution/node/316/log)
* [Report failed tests, if necessary](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/execution/node/675/log)

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/8cfbe189cfe29b8d94f5600cfe1be58956777ca8) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 267 | +9 | 258 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/8cfbe189cfe29b8d94f5600cfe1be58956777ca8) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 584.5KB | +6.9KB | 577.6KB |

### History
* :green_heart: [Build #62870](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/) succeeded 62a776642a632cfd6a8b26e7ec51036c8254767a
* :green_heart: [Build #62558](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/) succeeded 47484b0e90fc714a1980a38e0734228603cf96dd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/","number":"62558","commit":"47484b0e90fc714a1980a38e0734228603cf96dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/","number":"62870","commit":"62a776642a632cfd6a8b26e7ec51036c8254767a"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/","number":"62899","commit":"8cfbe189cfe29b8d94f5600cfe1be58956777ca8"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62943/)
* Commit: b0cdee06c42d30ad9a2ad1908eb55757310ec7e1

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/b0cdee06c42d30ad9a2ad1908eb55757310ec7e1) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 267 | +9 | 258 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/b0cdee06c42d30ad9a2ad1908eb55757310ec7e1) | diff | [baseline](https://github.com/elastic/kibana/commit/105e3a6c7ecc39442f10c8ec030f6297e0b56cd4) |
| --- | --- | --- | --- |
| ingestPipelines | 584.5KB | +6.9KB | 577.6KB |

### History
* :broken_heart: [Build #62899](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/) failed 8cfbe189cfe29b8d94f5600cfe1be58956777ca8
* :green_heart: [Build #62870](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/) succeeded 62a776642a632cfd6a8b26e7ec51036c8254767a
* :green_heart: [Build #62558](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/) succeeded 47484b0e90fc714a1980a38e0734228603cf96dd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62558/","number":"62558","commit":"47484b0e90fc714a1980a38e0734228603cf96dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62870/","number":"62870","commit":"62a776642a632cfd6a8b26e7ec51036c8254767a"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62899/","number":"62899","commit":"8cfbe189cfe29b8d94f5600cfe1be58956777ca8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/62943/","number":"62943","commit":"b0cdee06c42d30ad9a2ad1908eb55757310ec7e1"}]}
PIPELINE-->

cjcenizal commented:
@jloleysens I created https://github.com/elastic/kibana/issues/72568.
